// openclaw/dist/extensions/tavily/web-search-contract-api.js
import { i as buildTavilyWebSearchProviderBase, n as TAVILY_GENERIC_SEARCH_DESCRIPTION, r as TAVILY_GENERIC_SEARCH_SCHEMA } from "../../web-search-shared-Nc_uzO6c.js";
//#region extensions/tavily/web-search-contract-api.ts
let tavilySearchProviderModulePromise;
function loadTavilySearchProviderModule() {
	tavilySearchProviderModulePromise ??= import("../../tavily-search-provider-BiYXDdSW.js");
	return tavilySearchProviderModulePromise;
}
function createTavilyWebSearchProvider() {
	return {
		...buildTavilyWebSearchProviderBase(),
		createTool: (ctx) => ({
			description: TAVILY_GENERIC_SEARCH_DESCRIPTION,
			parameters: TAVILY_GENERIC_SEARCH_SCHEMA,
			execute: async (args) => {
				const { createTavilyWebSearchProvider: createRuntimeProvider } = await loadTavilySearchProviderModule();
				const tool = createRuntimeProvider().createTool(ctx);
				if (!tool) throw new Error("Tavily web_search provider did not create a runtime tool.");
				return await tool.execute(args);
			}
		})
	};
}
//#endregion
export { createTavilyWebSearchProvider };
