// openclaw/dist/codex-runtime-plugin-install-WVIUeSM1.js
import { a as modelSelectionShouldEnsureCodexPlugin } from "./openai-routing-DXJmS9CT.js";
import { t as createRuntimePluginModelSelectionHelpers } from "./runtime-plugin-install-CNFIPqvc.js";
//#region src/commands/codex-runtime-plugin-install.ts
const CODEX_RUNTIME_PLUGIN_ID = "codex";
const CODEX_RUNTIME_PLUGIN_LABEL = "Codex";
const CODEX_RUNTIME_PLUGIN_DESCRIPTOR = {
	pluginId: CODEX_RUNTIME_PLUGIN_ID,
	label: CODEX_RUNTIME_PLUGIN_LABEL,
	npmSpec: "@openclaw/codex",
	warningLabel: CODEX_RUNTIME_PLUGIN_LABEL
};
/** Return true when a selected model requires the Codex runtime plugin to be installed. */
function selectedModelShouldEnsureCodexRuntimePlugin(params) {
	return modelSelectionShouldEnsureCodexPlugin({
		config: params.cfg,
		model: params.model
	});
}
const codexRuntimePluginInstall = createRuntimePluginModelSelectionHelpers({
	descriptor: CODEX_RUNTIME_PLUGIN_DESCRIPTOR,
	shouldEnsure: selectedModelShouldEnsureCodexRuntimePlugin
});
const ensureCodexRuntimePluginForModelSelection = codexRuntimePluginInstall.ensure;
const repairCodexRuntimePluginInstallForModelSelection = codexRuntimePluginInstall.repair;
//#endregion
export { selectedModelShouldEnsureCodexRuntimePlugin as i, ensureCodexRuntimePluginForModelSelection as n, repairCodexRuntimePluginInstallForModelSelection as r, CODEX_RUNTIME_PLUGIN_ID as t };
