// openclaw/dist/models-CkzKfzm3.js
import { n as buildManifestModelProviderConfig } from "./provider-catalog-shared-CbHkdit7.js";
//#endregion
//#region extensions/gmi/models.ts
const GMI_MANIFEST_PROVIDER = buildManifestModelProviderConfig({
	providerId: "gmi",
	catalog: {
		"aliases": {
			"gmi-cloud": { "provider": "gmi" },
			"gmicloud": { "provider": "gmi" }
		},
		"providers": { "gmi": {
			"baseUrl": "https://api.gmi-serving.com/v1",
			"api": "openai-completions",
			"models": [
				{
					"id": "zai-org/GLM-5.1-FP8",
					"name": "GLM-5.1 FP8",
					"reasoning": true,
					"input": ["text"],
					"contextWindow": 202752,
					"maxTokens": 65536,
					"cost": {
						"input": 0,
						"output": 0,
						"cacheRead": 0,
						"cacheWrite": 0
					}
				},
				{
					"id": "deepseek-ai/DeepSeek-V3.2",
					"name": "DeepSeek V3.2",
					"reasoning": false,
					"input": ["text"],
					"contextWindow": 163840,
					"maxTokens": 65536,
					"cost": {
						"input": 0,
						"output": 0,
						"cacheRead": 0,
						"cacheWrite": 0
					}
				},
				{
					"id": "moonshotai/Kimi-K2.5",
					"name": "Kimi K2.5",
					"reasoning": true,
					"input": ["text", "image"],
					"contextWindow": 262144,
					"maxTokens": 65536,
					"cost": {
						"input": 0,
						"output": 0,
						"cacheRead": 0,
						"cacheWrite": 0
					}
				},
				{
					"id": "google/gemini-3.1-flash-lite",
					"name": "Gemini 3.1 Flash Lite",
					"reasoning": false,
					"input": ["text", "image"],
					"contextWindow": 1048576,
					"maxTokens": 65536,
					"cost": {
						"input": 0,
						"output": 0,
						"cacheRead": 0,
						"cacheWrite": 0
					}
				},
				{
					"id": "anthropic/claude-sonnet-4.6",
					"name": "Claude Sonnet 4.6",
					"reasoning": false,
					"input": ["text", "image"],
					"contextWindow": 2e5,
					"maxTokens": 64e3,
					"cost": {
						"input": 0,
						"output": 0,
						"cacheRead": 0,
						"cacheWrite": 0
					}
				},
				{
					"id": "openai/gpt-5.4",
					"name": "GPT-5.4",
					"reasoning": true,
					"input": ["text", "image"],
					"contextWindow": 4e5,
					"maxTokens": 128e3,
					"cost": {
						"input": 0,
						"output": 0,
						"cacheRead": 0,
						"cacheWrite": 0
					}
				}
			]
		} }
	}.providers.gmi
});
const GMI_BASE_URL = GMI_MANIFEST_PROVIDER.baseUrl;
const GMI_MODEL_CATALOG = GMI_MANIFEST_PROVIDER.models;
const GMI_DEFAULT_MODEL_REF = "gmi/google/gemini-3.1-flash-lite";
function buildGmiModelDefinition(model) {
	return {
		...model,
		api: "openai-completions"
	};
}
//#endregion
export { buildGmiModelDefinition as i, GMI_DEFAULT_MODEL_REF as n, GMI_MODEL_CATALOG as r, GMI_BASE_URL as t };
