// openclaw/dist/prompt-overlay-CMYIe9y9.js
import { i as GPT5_HEARTBEAT_PROMPT_OVERLAY, l as resolveGpt5SystemPromptContribution, s as renderGpt5PromptOverlay, t as GPT5_BEHAVIOR_CONTRACT } from "./gpt5-prompt-overlay-D4Aews98.js";
import "./provider-model-shared-DUdP4MPm.js";
//#region extensions/codex/prompt-overlay.ts
/**
* Codex prompt-overlay facade for GPT-5 behavior and heartbeat guidance.
*/
/** GPT-5 behavior contract re-exported under the Codex provider namespace. */
const CODEX_GPT5_BEHAVIOR_CONTRACT = GPT5_BEHAVIOR_CONTRACT;
/** Heartbeat prompt overlay re-exported under the Codex provider namespace. */
const CODEX_GPT5_HEARTBEAT_PROMPT_OVERLAY = GPT5_HEARTBEAT_PROMPT_OVERLAY;
/** Resolves the Codex system-prompt contribution for GPT-5-family models. */
function resolveCodexSystemPromptContribution(params) {
	return resolveGpt5SystemPromptContribution(params);
}
/** Renders the Codex prompt overlay text for supported GPT-5-family models. */
function renderCodexPromptOverlay(params) {
	return renderGpt5PromptOverlay(params);
}
//#endregion
export { resolveCodexSystemPromptContribution as i, CODEX_GPT5_HEARTBEAT_PROMPT_OVERLAY as n, renderCodexPromptOverlay as r, CODEX_GPT5_BEHAVIOR_CONTRACT as t };
