// openclaw/dist/sqlite-wal-DqkFpj0C.js
import { n as MAX_TIMER_TIMEOUT_MS } from "./number-coercion-CJQ8TR--.js";
import { i as formatErrorMessage } from "./errors-CbY2H91D.js";
import { t as installProcessWarningFilter } from "./warning-filter-p3Ue6g9S.js";
import "./number-coercion-Z7n6tXLk.js";
import { createRequire } from "node:module";
import fs from "node:fs";
import path from "node:path";
import childProcess from "node:child_process";
//#region src/infra/node-sqlite.ts
const require = createRequire(import.meta.url);
/** Load node:sqlite after installing the process warning filter. */
function requireNodeSqlite() {
	installProcessWarningFilter();
	try {
		return require("node:sqlite");
	} catch (err) {
		const message = formatErrorMessage(err);
		throw new Error(`SQLite support is unavailable in this Node runtime (missing node:sqlite). ${message}`, { cause: err });
	}
}
const LINUX_NFS_SUPER_MAGIC = 26985;
const PROC_MOUNTINFO_PATH = "/proc/self/mountinfo";
function normalizeNonNegativeInteger(value, label) {
	if (!Number.isInteger(value) || value < 0) throw new Error(`${label} must be a non-negative integer`);
	return value;
}
function findExistingVolumePath(targetPath) {
	let current = path.resolve(targetPath);
	while (true) try {
		return fs.statSync(current).isDirectory() ? current : path.dirname(current);
	} catch {
		const parent = path.dirname(current);
		if (parent === current) return null;
		current = parent;
	}
}
function decodeMountPath(value) {
	return value.replace(/\\([0-7]{3})/g, (_match, octal) => String.fromCharCode(Number.parseInt(octal, 8)));
}
function parseProcMountInfoEntries(contents) {
	const entries = [];
	for (const line of contents.split("\n")) {
		const separator = line.indexOf(" - ");
		if (separator === -1) continue;
		const fields = line.slice(0, separator).split(" ");
		const suffixFields = line.slice(separator + 3).split(" ");
		const mountPoint = fields[4];
		const fsType = suffixFields[0];
		if (mountPoint && fsType) entries.push({
			mountPoint: decodeMountPath(mountPoint),
			fsType
		});
	}
	return entries;
}
function parseMountCommandEntries(contents) {
	const entries = [];
	for (const line of contents.split("\n")) {
		const linuxMatch = /^.* on (.+) type ([^,\s)]+) \(/.exec(line);
		if (linuxMatch) {
			entries.push({
				mountPoint: linuxMatch[1],
				fsType: linuxMatch[2]
			});
			continue;
		}
		const bsdMatch = /^.* on (.+) \(([^,\s)]+)/.exec(line);
		if (bsdMatch) entries.push({
			mountPoint: bsdMatch[1],
			fsType: bsdMatch[2]
		});
	}
	return entries;
}
function readMountEntries() {
	try {
		return parseProcMountInfoEntries(fs.readFileSync(PROC_MOUNTINFO_PATH, "utf8"));
	} catch {}
	try {
		return parseMountCommandEntries(String(childProcess.execFileSync("mount", [])));
	} catch {
		return [];
	}
}
function isPathWithinMount(targetPath, mountPoint) {
	const resolvedTarget = path.resolve(targetPath);
	const resolvedMountPoint = path.resolve(mountPoint);
	return resolvedTarget === resolvedMountPoint || resolvedMountPoint === path.parse(resolvedMountPoint).root || resolvedTarget.startsWith(`${resolvedMountPoint}${path.sep}`);
}
function isNfsMountType(fsType) {
	return fsType.toLowerCase().startsWith("nfs");
}
function isNfsMountEntryPath(targetPath) {
	const mountEntry = readMountEntries().filter((entry) => isPathWithinMount(targetPath, entry.mountPoint)).toSorted((a, b) => b.mountPoint.length - a.mountPoint.length)[0];
	return mountEntry ? isNfsMountType(mountEntry.fsType) : false;
}
function isNfsBackedPath(targetPath) {
	if (typeof fs.statfsSync !== "function") return isNfsMountEntryPath(targetPath);
	const checkedPath = findExistingVolumePath(targetPath);
	if (!checkedPath) return false;
	try {
		if (fs.statfsSync(checkedPath).type === LINUX_NFS_SUPER_MAGIC) return true;
	} catch {
		return isNfsMountEntryPath(checkedPath);
	}
	return isNfsMountEntryPath(checkedPath);
}
function readJournalModeResult(row) {
	if (!row || typeof row !== "object") return null;
	const record = row;
	const value = record.journal_mode ?? Object.values(record)[0];
	return typeof value === "string" ? value.toLowerCase() : null;
}
function requireRollbackJournalMode(db, options) {
	const journalMode = readJournalModeResult(db.prepare("PRAGMA journal_mode = DELETE;").get());
	if (journalMode !== "delete") {
		const label = options.databaseLabel ?? "sqlite database";
		const location = options.databasePath ? ` at ${options.databasePath}` : "";
		throw new Error(`${label}${location} is on an NFS-backed volume but SQLite kept journal_mode=${journalMode ?? "unknown"}; refusing to continue with WAL on NFS.`);
	}
}
/** Configure WAL pragmas and return a handle for checkpoint/close maintenance. */
function configureSqliteWalMaintenance(db, options = {}) {
	const autoCheckpointPages = normalizeNonNegativeInteger(options.autoCheckpointPages ?? 1e3, "autoCheckpointPages");
	const checkpointIntervalMs = normalizeNonNegativeInteger(options.checkpointIntervalMs ?? 18e5, "checkpointIntervalMs");
	const timerIntervalMs = Math.min(checkpointIntervalMs, MAX_TIMER_TIMEOUT_MS);
	const checkpointMode = options.checkpointMode ?? "TRUNCATE";
	if (options.databasePath && isNfsBackedPath(options.databasePath)) {
		requireRollbackJournalMode(db, options);
		return {
			checkpoint: () => true,
			close: () => true
		};
	}
	db.exec("PRAGMA journal_mode = WAL;");
	db.exec(`PRAGMA wal_autocheckpoint = ${autoCheckpointPages};`);
	const checkpoint = () => {
		try {
			db.exec(`PRAGMA wal_checkpoint(${checkpointMode});`);
			return true;
		} catch (error) {
			options.onCheckpointError?.(error);
			return false;
		}
	};
	let timer = null;
	if (timerIntervalMs > 0) {
		timer = setInterval(checkpoint, timerIntervalMs);
		timer.unref?.();
	}
	return {
		checkpoint,
		close: () => {
			if (timer) {
				clearInterval(timer);
				timer = null;
			}
			return checkpoint();
		}
	};
}
//#endregion
export { requireNodeSqlite as n, configureSqliteWalMaintenance as t };
