// openclaw/dist/tavily-search-provider-LMWzqJg-.js
import { g as readPositiveIntegerParam } from "./common-CEBsSh4n.js";
import "./param-readers-BqIX58yY.js";
import { i as buildTavilyWebSearchProviderBase, n as TAVILY_GENERIC_SEARCH_DESCRIPTION, r as TAVILY_GENERIC_SEARCH_SCHEMA } from "./web-search-shared-Nc_uzO6c.js";
//#region extensions/tavily/src/tavily-search-provider.ts
let tavilyClientModulePromise;
function loadTavilyClientModule() {
	tavilyClientModulePromise ??= import("./tavily-client-NRCibp4j.js");
	return tavilyClientModulePromise;
}
function createTavilyWebSearchProvider() {
	return {
		...buildTavilyWebSearchProviderBase(),
		createTool: (ctx) => ({
			description: TAVILY_GENERIC_SEARCH_DESCRIPTION,
			parameters: TAVILY_GENERIC_SEARCH_SCHEMA,
			execute: async (args) => {
				const { runTavilySearch } = await loadTavilyClientModule();
				return await runTavilySearch({
					cfg: ctx.config,
					query: typeof args.query === "string" ? args.query : "",
					maxResults: readPositiveIntegerParam(args, "count", {
						message: "count must be an integer from 1 to 20",
						max: 20
					})
				});
			}
		})
	};
}
//#endregion
export { createTavilyWebSearchProvider as t };
